### ISSUE TITLE
WebAssembly speed up the execution time of the web program

### ISSUE THREAD

**Issue by tclin914** [WebAssembly/design#528]
Hi,

I did a experiment that two bubble sort programs running in d8 the one written in JavaScript and the other one written in WebAssembly.The execution time of the javascript program is 4582.631ms, the WebAssembly program is 5567.51 ms (by performance.now()). The result is amazing for me. I expect that WebAssembly program should run fast than javascript program. I consider one of the reason may be bubble sort do not use any feature of the dynamic language, which is a cost for javascript.

But I really want to understand the compute performance between JavaScript and WebAssembly.

Thank you.

**Comment by jfbastien** [comment]
WebAssembly is still early, you may be hitting something that we haven't put optimization effort yet. Do you have code and repro instructions?

**Comment by jcbeyler** [comment]
Yes I agree entirely. There are too many variables here to understand what is going on; how was the Wasm generated, what does the JavaScript look like; what version of d8 were you using, what was the environment, was there noise, etc.?

I added a micro here:
https://github.com/WebAssembly/wasm-to-llvm-prototype/commit/5f85c586ab36f05ffb8d639e5fbf5c7dd3a5714e

Which does what you were saying on a small array and I got as results:
Wasm: Time between in average is 68048248.000000
Gcc-O3: Time between in average is 66598763.000000
Clang-O3: Time between in average is 39029055.000000

Which is the fifth set and sort of an array of 10k elements, set to be in inverse order (10000, 9999, 9998...).

This is to be taken with a lot of gloves since I did not look into it at all but it seems Wasm gets similar results as Gcc, Clang fairs better in this case...

Conclusion: In this case, my wasm version of bubble sort seems to fair as well as my C-version of bubble sort when using Gcc for the C part; it seems to fair twice as slowly when you look at clang.

(note the seems since I have not double checked this data... :) )

But you cannot look at one example (bubble sort) and make any judgement in general... There are way too many factors and even were you to say that Clang is better than Gcc in this case (which I am not saying because I've said that I don't automatically trust my numbers), you would still not be able to say that in the general case ;)
